File: Striver/Recursion/q7.py
# Reverse an array using recursion

# reversing array using 2 pointer approach recursively
def reverseArray(arr, start, end):
    if start >= end:
        return
    arr[start], arr[end] = arr[end], arr[start]
    reverseArray(arr, start + 1, end - 1)

# reversing array using single parameter recursively
def reverseArray2(arr, n):
    if n >= len(arr) // 2:
        return
    arr[n], arr[len(arr) - n - 1] = arr[len(arr) - n - 1], arr[n]
    reverseArray2(arr, n + 1)

arr = [1, 2, 3, 4, 5, 6]
print(arr)
reverseArray(arr, 0, 5)
print(arr)

File: Striver/Recursion/q7.java
package Striver.Recursion;

public class q7 {
    // Reverse an array using recursion
    public static void main(String[] args) {
        int[] arr = { 1, 2, 3, 4, 5, 6, 7, 8, 9 };
        reverse(arr, 0, arr.length - 1);
        for (int i : arr) {
            System.out.print(i + " ");
        }
    }

    //reversing array by 2 pointer method recusively
    public static void reverse(int[] arr, int start, int end) {
        if (start >= end) {
            return;
        }
        int temp = arr[start];
        arr[start] = arr[end];
        arr[end] = temp;
        reverse(arr, start + 1, end - 1);
    }

    //reversing array by passing single parameter recusively
    public static void reverse2(int[] arr, int start) {
        if (start >= arr.length / 2) {
            return;
        }
        int temp = arr[start];
        arr[start] = arr[arr.length - 1 - start];
        arr[arr.length - 1 - start] = temp;
        reverse2(arr, start + 1);
    }
}
